### ISSUE TITLE
Can't upload large networks

### ISSUE THREAD

**Issue by gcp** [gcp/leela-zero-server#23]
Uploading the 256x20 network from the https://github.com/gcp/leela-zero/issues/884 keeps failing.

```
curl -F'key=blah' -F'weights=@leelaz-model-3008000.txt.gz' -F'training_steps=3008000' -F'training_count=0' http://zero.sjeng.org/submit-network

<!DOCTYPE html>
<html>
<head>
<title>Error</title>
<style>
    body {
        width: 35em;
        margin: 0 auto;
        font-family: Tahoma, Verdana, Arial, sans-serif;
    }
</style>
</head>
<body>
<h1>An error occurred.</h1>
<p>Sorry, the page you are looking for is currently unavailable.<br/>
Please try again later.</p>
<p>If you are the system administrator of this resource then you should check
the <a href="http://nginx.org/r/error_log">error log</a> for details.</p>
<p><em>Faithfully yours, nginx.</em></p>
</body>
</html>
```

I tried increasing all relevant nginx configuration (size, timeout), no dice.

I tried ssh'ing the file over, and doing the upload on the machine itself (which should avoid transmission speed issues). It's still very slow, and eventually gives the same error.

So I think the issue is at the node level.

**Comment by roy7** [comment]
Try it again when you have time. I turned off the redirect of stderr to stdout so if there are any node level error messages I should see them on the console.

**Comment by roy7** [comment]
A tool called Multer might be of help to us. https://stackoverflow.com/questions/40723536/uploading-large-file-to-nodejs-webserver

**Comment by roy7** [comment]
Possibly related.

https://github.com/AnalyticalGraphicsInc/obj2gltf/issues/59

**Comment by roy7** [comment]
Real solution is probably to move to using streams. Stream the upload to disk. Stream from disk into hash function. Never store full network file in ram?

https://carlosrymer.com/using-node-streams-to-handle-large-file-uploads-24c1a0141b9c

**Comment by roy7** [comment]
What does the nginx error log say?

**Comment by gcp** [comment]
```
2018/03/06 11:24:01 [error] 6208#6208: *29887681 upstream prematurely closed connection while reading response heade
r from upstream, client: my_ip_here, server: zero.sjeng.org, request: "POST /submit-netw
ork HTTP/1.1", upstream: "http://127.0.0.1:8080/submit-network", host: "zero.sjeng.org"
```

**Comment by roy7** [comment]
If streams don't do what we want we could read a file in chunks and calculate the hash incrementally.

https://github.com/satazor/js-spark-md5

**Comment by roy7** [comment]
I don't know why this is failing yet. Need to test some. It wasn't crashing the server as a whole right? So maybe not unzip/ram related problems. I wonder if there's a max upload size we're violating. Worst case we could add a new endpoint to 'import network' which will read a network file off the disk instead of via web upload. Then you change your side to scp and import command, instead of a single submit command.

**Comment by rchs819** [comment]
This guy had same error message.
https://stackoverflow.com/questions/36488688/ngnix-upstream-prematurely-closed-connection-while-reading-response-header-from

The solution is to increase [server.timeout](https://nodejs.org/api/http.html#http_server_timeout) or setting it to `0` to disable timeout.
The default server timeout is 2 min, @gcp do you recall how long did it take to upload the large network?

**Comment by gcp** [comment]
I did another test by uploading the data to the server, and injecting directly into node.js (bypassing nginx). This fails, so I'm sure it's not my nginx configuration at fault.

**Comment by dtracers** [comment]
Instead of uploading the entire network as one file you could split the weights into 2 parts then merge them at the end

**Comment by gcp** [comment]
This isn't implemented on the server.

**Comment by gcp** [comment]
It looks like this is basically a stupid Chrome limitation: https://bugs.chromium.org/p/v8/issues/detail?id=6148

Anything V8 based (including node) can't handle strings >256M.

The 256x20 files are like ~270M unzipped.

**Comment by gcp** [comment]
V8 6.2.10 is supposed to have the fix, and it seems node v8.10 has 6.2.414.50. I updated, maybe I just need to kick the server?

**Comment by roy7** [comment]
Too easy!
